### ISSUE TITLE
Feature request: add a tag for the unit test design

### ISSUE THREAD

**Issue by drgorillamd** [alexfertel/bulloak#65]
After listing all branches of a control flow, multiple approach/testing strategies are possible, leading to different tree and test suite complexity (eg branch testing, basis and full path testing, boundary cases).

It would be interesting to find a way to label the underlying design of a tree (and reflect it in the test contract name or top-level comment).

 For instance:
```solidity
contract TestMe {
    function foo(bool a, bool b) external returns (bool) {
        if(a) emit EventA();
        if(b) emit EventB();
    }
}
```

branch coverage
```
TestMe::foo(branch)
├── when a is true
│   └── if should emit  EventA
└── when b is true
    └── it should emit EventB
```
path coverage
```
TestMe::foo(full path)
├── given a is true
│   ├── when b is true
│   │   ├── it should emit EventA
│   │   └── it should emit EventB
│   └── when b is false
│       └── it should emit EventA
└── given a is false
    ├── when b is true
    │   └── it should emit EventB
    └── when b is false
        └── it should not emit any event
```

would then scaffold 2 tests contracts, `TestMe_Test_Branch` and `TestMe_Test_FullPath`

**Comment by alexfertel** [comment]
I see, that sounds super interesting. It's a tiny bit big in scope, but I think it's reasonable.

Could you give a more practical example? If I understand correctly, `full path` contains `branch`, so it's not that useful.

Also, it would be nice to get @PaulRBerg 's opinion as well.

**Comment by alexfertel** [comment]
> not always doable (see [inf.ed.ac.uk/teaching/courses/st/2017-18/Path-coverage.pdf](https://www.inf.ed.ac.uk/teaching/courses/st/2017-18/Path-coverage.pdf) for a pretty cool example)

Ah, this is great, now I see it clearly, thanks for sharing!

This is really, really useful, so we should actually prioritize it. I think we need to design a bit how it will look like though, not convinced that emitting different contracts is the way.

**Comment by alexfertel** [comment]
Just to be clear, having two different trees for the same function in the same `.tree` is not enough for this use case?

```
TestMe::foo // branch
├── when a and b are true
│   └── it should return true
└── when a is false
    └── it should return false
   
TestMe::foo // Full path
├── given a is true
│   ├── when b is true
│   │   └── it should return true
│   └── when b is false
│       └── it should return false
└── given a is false
    ├── when b is true
    │   └── it should return false
    └── when b is false
        └── it should return false
```

generates

```solidity
contract TestMe {
    function test_FooWhenAAndBAreTrue() external {
        // it should return true
    }

    function test_FooWhenAIsFalse() external {
        // it should return false
    }

    modifier givenAIsTrue() {
        _;
    }

    function test_FooWhenBIsTrue() external givenAIsTrue {
        // it should return true
    }

    function test_FooWhenBIsFalse() external givenAIsTrue {
        // it should return false
    }

    modifier givenAIsFalse() {
        _;
    }

    function test_FooWhenBIsTrue2() external givenAIsFalse {
        // it should return false
    }

    function test_FooWhenBIsFalse2() external givenAIsFalse {
        // it should return false
    }
}
```

**Comment by drgorillamd** [comment]
You're right, it would be for this example, but in real use-case, one usually pick a single coverage (so this would rather pin the difference between different functions tested using different coverage -[ random example in the /trees/ files of this pr](https://github.com/Bananapus/nana-swap-terminal/pull/10/files)). 

 For instance, if you're testing a contract with 3 external functions, maybe 2 of them are super short/simple flow (and a path coverage is realistic) while the third one has a huge number of branches and conditional op (so branch coverage is probably better), imo, this difference should be documented somewhere.
I tend to write natspecs for my test function in forge, and add it there - maybe having the equivalent while scaffolding would make more sense (eg any comment at the root level or just above in the .tree is included in the test function natspecs as @dev for instance)

**Comment by drgorillamd** [comment]
NB this could be used for other test "tags" too actually (fuzz vs static for instance), so a tree comment which is translated to a `/// @dev` makes probably more sense

**Comment by alexfertel** [comment]
>  imo, this difference should be documented somewhere.

I see, I think this is the key point, right? It shouldn't be surprising to people reading which kind of coverage they are using.

I agree this should be clear.

An interesting thing to think about is how does the test evolve with the evolution of the function. Say that we used branch coverage initially for this function:

```solidity
bool foo (bool a, bool b) {
    if (a && b) {
        return true;
    }
    return false;
}
```

```
TestMe::foo(branch)
├── when a and b are true
│   └── it should return true
└── when a is false
    └── it should return false
```

(which note that is not an entirely correct spec, because the second branch should be something like "Otherwise")

But then we change it to:

```solidity
bool foo (bool a, bool b) {
    if (a) {
      if (b) {
	    return true;
      }
      return false;
    }
    else {
      if (b) {
        return false;
      }
      return true;
    }
}
```

(this function is just `a == b`, but it's a dumb example, bear with me)

Now path coverage is the easier way to maintain the spec.

> NB this could be used for other test "tags" too actually (fuzz vs static for instance)

Could you expand a bit with an example of this? Sorry about all the back and forth, I want to make sure we get this right.

**Comment by PaulRBerg** [comment]
unfortunately not able to revieswthis proposal at this time, tagging @andreivladbrg and @smol-ninja in case they want to contribute.

**Comment by alexfertel** [comment]
Sry for the delayed response!

> _parsing the comments after a root or a when/given in the .tree file and emit them as @dev natspecs (for the contract or the test function). What do you think @alexfertel ? This could include any coverage or extra-info, replicated in both files

Hmmm, could you give a practical example of how this would look like?

**Comment by drgorillamd** [comment]
Sure! In other words, it's including the comments as additional `dev` natspecs (except for the `it`, as they're already translated to comments)
```
TestMe::foo // This is a branch coverage.
├── when a and b are true
│   └── it should return true
└── when a is false // b false would cover the same branch
    └── it should return false
```
would generate
```solidity
/// @dev This is a branch coverage
contract TestMe {
    function test_FooWhenAAndBAreTrue() external {
        // it should return true
    }

    /// @dev b false would cover the same branch
    function test_FooWhenAIsFalse() external {
        // it should return false
    }
}
```

**Comment by alexfertel** [comment]
Cool, that makes sense 👍 

The example is not accurate though, because `it`s are what is mapped 1 to 1 with test functions. Consider this dummy example:

```
TestMe::foo // This is a branch coverage.
└── when a is false // first comment.
	└── when b is false // second comment.
    	└── it should return false
```

which generates

```solidity
contract TestMefoo {
    modifier whenAIsFalse() {
        _;
    }

    function test_WhenBIsFalse() external whenAIsFalse {
        // it should return false
    }
}
```

Which comment should become the natspec?

**Comment by drgorillamd** [comment]
I think this should become: 
```solidity
/// @dev This is a branch coverage.
contract TestMefoo {

    /// @dev first comment.
    modifier whenAIsFalse() {
        _;
    }

    /// @dev second comment.
    function test_WhenBIsFalse() external whenAIsFalse {
        // it should return false
    }
}
```
wdyt?

**Comment by alexfertel** [comment]
Sgtm, let's do it!

**Comment by alexfertel** [comment]
I think this doesn't need a CLI flag since it's additive and makes sense.

I'd rather we don't emit `@dev` because that way we can support users who don't want the `@dev` in the emitted comments. This is debatable though, if we want to be opinionated, that's fine too! Lmk what you think.

**Comment by alexfertel** [comment]
As always, if you need any help lmk. If you don't plan to work on it lmk as well.

**Comment by drgorillamd** [comment]
Hmm, true, natspecs only makes sense if docs are generated, which is perhaps not routinely done.
I'll give it a go soon, few other deadlines before (and the 2 other issues I've opened;) Thx for the help!

**Comment by alexfertel** [comment]
It's looking good! Parsing is the roughest part of the code, so good luck! 🙈
